Issue #12806 - Overhaul ComplianceViolation.Listener reporting

joakime opened the pull request:
Cleanup all code paths that perform a ComplianceViolation check to do the following ...

* Always report the violation, if detected.
* Report if the violation was allowed (or not)
* Use the `HttpChannelState.getCompliaceViolationListener()` in all cases (this gets the benefit of a single `ComplianceViolation.Listener` to use, along with all of the `initialize()`, `onRequestBegin()`, and `onRequestEnd()` techniques.
* Deprecate the `HttpConfiguration` notify methods
* Deprecate the `ComplianceViolation.notify()` static methods.

Closes #12806

joakime committed:
Issue #12806 - Overhaul ComplianceViolation.Listener reporting

joakime committed:
Fixing HttpParserTest failures

joakime committed:
Fixing RFC6265CookieParserTest failures

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpQuotedCSV.java:
```diff
@@ -0,0 +1,68 @@
+
+package org.eclipse.jetty.http;
+
+public class HttpQuotedCSV extends QuotedCSV
```
javadoc definitely needed to explain the difference!

Edit 1:
Plus I don't think the name is correct.  Make it a subtype: `QuotedCSV.Compliant`

Edit 2:
Oh it was already a subtype?  Why a top level class for such a trivial addition?  At  the very least keep the name! You can see how confused I was by this class!

joakime replied:
This needs ComplianceViolation.Listener, which exists in jetty-http, and the concept is jetty-http specific.

The BiConsumer hack doesn't work anymore, starting with this PR.

joakime replied:
This needs ComplianceViolation.Listener, which exists in jetty-http, and the concept is jetty-http specific.

The BiConsumer hack doesn't work anymore, starting with this PR.

gregw replied:
But QuotedCSV and QuotedQualityCSV are both in `jetty-http`, so there is no visibility issue and no need to change the name or location.    It is easy to replace the BiConsumer:
```diff
diff --git a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
index d45fac8b882..4f431bd751b 100644
--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java
@@ -150,7 +150,7 @@ public interface HttpFields extends Iterable<HttpField>, Supplier<HttpFields>
         return new org.eclipse.jetty.http.MutableHttpFields(fields, removeFields);
     }
 
-    static Mutable build(HttpCompliance httpCompliance, BiConsumer<ComplianceViolation, String> notifyViolation)
+    static Mutable build(HttpCompliance httpCompliance, ComplianceViolation.Listener notifyViolation)
     {
         return new org.eclipse.jetty.http.MutableHttpFields.Compliant(httpCompliance, notifyViolation);
     }
diff --git a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
index 28f168cac8c..20117af322a 100644
--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
@@ -2360,22 +2360,14 @@ public class HttpParser
         protected QuotedCSV newQuotedCSV(boolean keepQuotes, String value)
         {
             if (getHeader() != null && HttpField.ETAG_HEADER.contains(this.getHeader()))
-                return new QuotedCSV.Etags(_complianceMode,
-                    (v, r) ->
+                return new QuotedCSV.Etags(_complianceMode, new ComplianceViolation.Listener()
+                {
+                    @Override
+                    public void onComplianceViolation(ComplianceViolation.Event event)
                     {
-                        try
-                        {
-                            _handler.onViolation(new ComplianceViolation.Event(_complianceMode, v, r));
-                        }
-                        catch (BadMessageException bme)
-                        {
-                            throw bme;
-                        }
-                        catch (Throwable t)
-                        {
-                            throw new BadMessageException(t.getMessage(), t);
-                        }
-                    }, value);
+                        _handler.onViolation(event);
+                    }
+                }, value);
 
             return new QuotedCSV(keepQuotes, value)
             {
diff --git a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MutableHttpFields.java b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MutableHttpFields.java
index fbe20df2419..6dde72b7027 100644
--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MutableHttpFields.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MutableHttpFields.java
@@ -596,9 +596,9 @@ class MutableHttpFields implements HttpFields.Mutable
     public static class Compliant extends org.eclipse.jetty.http.MutableHttpFields
     {
         private final HttpCompliance _httpCompliance;
-        private final BiConsumer<ComplianceViolation, String> _notifyViolation;
+        private final ComplianceViolation.Listener _notifyViolation;
 
-        public Compliant(HttpCompliance httpCompliance, BiConsumer<ComplianceViolation, String> notifyViolation)
+        public Compliant(HttpCompliance httpCompliance, ComplianceViolation.Listener notifyViolation)
         {
             _httpCompliance = httpCompliance;
             _notifyViolation = notifyViolation;
diff --git a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSV.java b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSV.java
index 0a70df56daa..40be865745c 100644
--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSV.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSV.java
@@ -197,14 +197,14 @@ public class QuotedCSV extends QuotedCSVParser implements Iterable<String>
     public static class Compliant extends QuotedCSV
     {
         private final ComplianceViolation.Mode _complianceMode;
-        private final BiConsumer<ComplianceViolation, String> _violationNotifier;
+        private final ComplianceViolation.Listener _violationNotifier;
 
-        public Compliant(ComplianceViolation.Mode complianceMode, BiConsumer<ComplianceViolation, String> violationNotifier)
+        public Compliant(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener violationNotifier)
         {
             this(complianceMode, violationNotifier, true);
         }
 
-        public Compliant(ComplianceViolation.Mode complianceMode, BiConsumer<ComplianceViolation, String> violationNotifier, boolean keepQuotes, String... values)
+        public Compliant(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener violationNotifier, boolean keepQuotes, String... values)
         {
             super(keepQuotes, values);
             _complianceMode = complianceMode;
@@ -215,7 +215,7 @@ public class QuotedCSV extends QuotedCSVParser implements Iterable<String>
         protected void onComplianceViolation(ComplianceViolation violation, String value)
         {
             if (_complianceMode != null && _complianceMode.allows(violation))
-                _violationNotifier.accept(violation, value);
+                _violationNotifier.onComplianceViolation(new ComplianceViolation.Event(_complianceMode, violation, value));
             else
                 super.onComplianceViolation(violation, value);
         }
@@ -223,7 +223,7 @@ public class QuotedCSV extends QuotedCSVParser implements Iterable<String>
 
     public static class Etags extends Compliant
     {
-        public Etags(ComplianceViolation.Mode complianceMode, BiConsumer<ComplianceViolation, String> violationNotifier, String... values)
+        public Etags(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener violationNotifier, String... values)
         {
             super(complianceMode, violationNotifier, true, values);
         }
diff --git a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java
index f46d1937fdd..d17c653259e 100644
--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java
@@ -18,7 +18,6 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
-import java.util.function.BiConsumer;
 import java.util.function.ToIntFunction;
 import java.util.stream.Collectors;
 
@@ -330,23 +329,23 @@ public class QuotedQualityCSV extends QuotedCSV implements Iterable<String>
     public static class Compliant extends QuotedQualityCSV
     {
         private final ComplianceViolation.Mode _complianceMode;
-        private final BiConsumer<ComplianceViolation, String> _violationNotifier;
+        private final ComplianceViolation.Listener _violationNotifier;
 
-        public Compliant(ComplianceViolation.Mode complianceMode, BiConsumer<ComplianceViolation, String> violationNotifier, String[] preferredOrder)
+        public Compliant(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener violationNotifier, String[] preferredOrder)
         {
             super(preferredOrder);
             _complianceMode = complianceMode;
             _violationNotifier = violationNotifier;
         }
 
-        public Compliant(ComplianceViolation.Mode complianceMode, BiConsumer<ComplianceViolation, String> violationNotifier, List<String> preferredOrder)
+        public Compliant(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener violationNotifier, List<String> preferredOrder)
         {
             super(preferredOrder);
             _complianceMode = complianceMode;
             _violationNotifier = violationNotifier;
         }
 
-        public Compliant(ComplianceViolation.Mode complianceMode, BiConsumer<ComplianceViolation, String> violationNotifier, ToIntFunction<String> secondaryOrdering)
+        public Compliant(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener violationNotifier, ToIntFunction<String> secondaryOrdering)
         {
             super(secondaryOrdering);
             _complianceMode = complianceMode;
@@ -357,7 +356,7 @@ public class QuotedQualityCSV extends QuotedCSV implements Iterable<String>
         protected void onComplianceViolation(ComplianceViolation violation, String value)
         {
             if (_complianceMode != null && _complianceMode.allows(violation))
-                _violationNotifier.accept(violation, value);
+                _violationNotifier.onComplianceViolation(new ComplianceViolation.Event(_complianceMode, violation, value));
             else
                 super.onComplianceViolation(violation, value);
         }
diff --git a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java
index 6ee09fac938..21d2b5bad89 100644
--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java
@@ -29,6 +29,7 @@ import org.eclipse.jetty.http.HttpMethod;
 import org.eclipse.jetty.http.HttpScheme;
 import org.eclipse.jetty.http.MultiPartCompliance;
 import org.eclipse.jetty.http.UriCompliance;
+import org.eclipse.jetty.util.Attributes;
 import org.eclipse.jetty.util.HostPort;
 import org.eclipse.jetty.util.Index;
 import org.eclipse.jetty.util.Jetty;
@@ -50,7 +51,7 @@ import org.eclipse.jetty.util.component.DumpableCollection;
  * </p>
  */
 @ManagedObject("HTTP Configuration")
-public class HttpConfiguration implements Dumpable
+public class HttpConfiguration implements Dumpable, ComplianceViolation.Listener
 {
     public static final String SERVER_VERSION = "Jetty(" + Jetty.VERSION + ")";
     private final List<Customizer> _customizers = new CopyOnWriteArrayList<>();
@@ -363,16 +364,17 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Set if true, delays the application dispatch until content is available (defaults to true).
+     *
      * @param delay if true, delays the application dispatch until content is available (defaults to true)
      * @deprecated Use {@link org.eclipse.jetty.server.handler.EagerContentHandler} instead.
      */
-    @Deprecated (forRemoval = true, since = "12.1.0")
+    @Deprecated(forRemoval = true, since = "12.1.0")
     public void setDelayDispatchUntilContent(boolean delay)
     {
         _delayDispatchUntilContent = delay;
     }
 
-    @Deprecated (forRemoval = true, since = "12.1.0")
+    @Deprecated(forRemoval = true, since = "12.1.0")
     public boolean isDelayDispatchUntilContent()
     {
         return _delayDispatchUntilContent;
@@ -380,6 +382,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Set whether to use direct ByteBuffers for reading.
+     *
      * @param useInputDirectByteBuffers whether to use direct ByteBuffers for reading
      */
     public void setUseInputDirectByteBuffers(boolean useInputDirectByteBuffers)
@@ -395,6 +398,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Set whether to use direct ByteBuffers for writing.
+     *
      * @param useOutputDirectByteBuffers whether to use direct ByteBuffers for writing
      */
     public void setUseOutputDirectByteBuffers(boolean useOutputDirectByteBuffers)
@@ -424,6 +428,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Set the size of the buffer into which HTTP request are read.
+     *
      * @param inputBufferSize buffer size in bytes.
      */
     public void setInputBufferSize(int inputBufferSize)
@@ -626,8 +631,8 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * @param minInputBufferSpace The minimum space available in a retained input buffer before allocating a new one;
-     *                            0 to always allocate a new buffer;
-     *                            -1 for a default value
+     * 0 to always allocate a new buffer;
+     * -1 for a default value
      */
     public void setMinInputBufferSpace(int minInputBufferSpace)
     {
@@ -763,6 +768,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Add a {@link ComplianceViolation.Listener} to the configuration
+     *
      * @param listener the listener to add
      */
     public void addComplianceViolationListener(ComplianceViolation.Listener listener)
@@ -772,6 +778,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Remove a {@link ComplianceViolation.Listener} from the configuration
+     *
      * @param listener the listener to remove
      * @return {@code true} if this list contained the specified element
      */
@@ -782,6 +789,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Get the list of configured {@link ComplianceViolation.Listener} to use.
+     *
      * @return the list of configured listeners
      */
     public List<ComplianceViolation.Listener> getComplianceViolationListeners()
@@ -789,6 +797,36 @@ public class HttpConfiguration implements Dumpable
         return this._complianceViolationListeners;
     }
 
+    @Override
+    public void onComplianceViolation(ComplianceViolation.Event event)
+    {
+        for (ComplianceViolation.Listener listener : getComplianceViolationListeners())
+        {
+            if (listener != this)
+                listener.onComplianceViolation(event);
+        }
+    }
+
+    @Override
+    public void onRequestBegin(Attributes request)
+    {
+        for (ComplianceViolation.Listener listener : getComplianceViolationListeners())
+        {
+            if (listener != this)
+                listener.onRequestBegin(request);
+        }
+    }
+
+    @Override
+    public void onRequestEnd(Attributes request)
+    {
+        for (ComplianceViolation.Listener listener : getComplianceViolationListeners())
+        {
+            if (listener != this)
+                listener.onRequestEnd(request);
+        }
+    }
+
     /**
      * Utility to notify a violation for the appropriate compliance mode
      *
@@ -809,6 +847,7 @@ public class HttpConfiguration implements Dumpable
 
     /**
      * Set whether remote errors, when detected, are notified to async applications.
+     *
      * @param notifyRemoteAsyncErrors whether remote errors, when detected, are notified to async applications
      */
     public void setNotifyRemoteAsyncErrors(boolean notifyRemoteAsyncErrors)
diff --git a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java
index 150fdfb8186..996ddafbde8 100644
--- a/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java
+++ b/jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java
@@ -449,7 +449,7 @@ public class ResourceService
         //   If-Match: W/"abc,xyz", "123456"
         // This means we have to parse with QuotedCSV all the time, as we cannot just
         // test for the existence of a "," (comma) in the value to know if it's delimited or not
-        QuotedCSV quoted = new QuotedCSV.Etags(httpConfiguration.getHttpCompliance(), httpConfiguration::notifyViolation, requestEtag);
+        QuotedCSV quoted = new QuotedCSV.Etags(httpConfiguration.getHttpCompliance(), httpConfiguration, requestEtag);
         for (String tag : quoted)
         {
             if (EtagUtils.matches(contentETag, tag))
diff --git a/jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceService.java b/jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceService.java
index b78590f6e8d..159e79c29e7 100644
--- a/jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceService.java
+++ b/jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceService.java
@@ -24,7 +24,6 @@ import java.util.Enumeration;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.BiConsumer;
 import java.util.function.Supplier;
 
 import jakarta.servlet.AsyncContext;
@@ -548,7 +547,6 @@ public class ResourceService
             if (_etags)
             {
                 ComplianceViolation.Mode mode = httpConfiguration == null ? null : httpConfiguration.getHttpCompliance();
-                BiConsumer<ComplianceViolation, String> notify = httpConfiguration == null ? null : httpConfiguration::notifyViolation;
 
                 String etag = content.getETagValue();
                 if (ifm != null)
@@ -556,7 +554,7 @@ public class ResourceService
                     boolean match = false;
                     if (etag != null)
                     {
-                        QuotedCSV quoted = new QuotedCSV.Etags(mode, notify, ifm);
+                        QuotedCSV quoted = new QuotedCSV.Etags(mode, httpConfiguration, ifm);
                         for (String etagWithSuffix : quoted)
                         {
                             if (EtagUtils.matches(etag, etagWithSuffix))
@@ -585,7 +583,7 @@ public class ResourceService
 
                     // Handle list of tags
 
-                    QuotedCSV quoted = new QuotedCSV.Etags(mode, notify, ifnm);
+                    QuotedCSV quoted = new QuotedCSV.Etags(mode, httpConfiguration, ifnm);
                     for (String tag : quoted)
                     {
                         if (EtagUtils.matches(etag, tag))
```

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpQuotedQualityCSV.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+import java.util.function.ToIntFunction;
+
+public class HttpQuotedQualityCSV extends QuotedQualityCSV
```
javadoc needed to explain what is this new class?

Edit1:
Also the name is not good.  I think do what we did with `HttpField` and make this a subtype: `QuotedQualityCSV.Compliant`

Edit2:
See comment on HttpQuotedCSV

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpQuotedQualityCSV.java:
```diff
@@ -0,0 +1,53 @@
+    private final ComplianceViolation.Mode complianceMode;
+    private final ComplianceViolation.Listener listener;
+
+    public HttpQuotedQualityCSV(ComplianceViolation.Mode complianceMode, ComplianceViolation.Listener listener, String[] preferredOrder)
```
I think we tend to put the super class arguments first?

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -381,6 +382,41 @@ public HttpCompliance without(String name, Violation... violations)
+
+        if (!allowed)
+            throw error.apply(violation.getDescription());
+    }
```
This method should be written against ComplianceViolation and be a static method somewhere rather than a repeated implementation in every violation subtype.

joakime replied:
Then it becomes generic with the ComplianceViolation.Mode and needs a instance of check for every type.
Also it does handle the 3 UriCompliance configs properly (request, response, redirect).

gregw replied:
why does it need an instance check?  the Event class takes a pure ComplianceViolation and not a specific type.

joakime committed:
Revert HttpQuotedCSV and HttpQuotedQualityCSV change.

joakime committed:
Call HttpChannelState.initialize()

joakime committed:
HttpFields & UrlEncoded updates

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieCache.java:
```diff
@@ -164,7 +164,10 @@ public List<HttpCookie> getCookies(HttpFields headers, ComplianceViolation.Liste
 
         if (_violations != null && !_violations.isEmpty())
+        {
+            CookieCompliance cookieCompliance = _parser.getCookieCompliance();
```
This variable is not used

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieCache.java:
```diff
@@ -164,7 +164,10 @@ public List<HttpCookie> getCookies(HttpFields headers, ComplianceViolation.Liste
         if (_violations != null && !_violations.isEmpty())
+        {
+            CookieCompliance cookieCompliance = _parser.getCookieCompliance();
             _violations.forEach(complianceViolationListener::onComplianceViolation);
```
should we assertNotNull on the cVL? or skip this line if it is null?

joakime replied:
The expectation is that there will always be a cVL, even if its just `ComplianceViolation.Listener.NOOP`.
I added a requireNonNull to the start of the method anyway.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CookieParser.java:
```diff
@@ -52,6 +52,8 @@ default void parseFields(List<String> rawFields) throws InvalidCookieException
             parseField(field);
     }
 
+    CookieCompliance getCookieCompliance();
```
javadoc

joakime replied:
Easier to just get rid of this method, it's no longer being used anyway.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -130,7 +131,13 @@ default Runnable onClose()
      */
     void initialize();
 
+    boolean complianceAllows(ComplianceViolation violation, String detail);
```
never called?

joakime replied:
Removed

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -130,7 +131,13 @@ default Runnable onClose()
 
+    boolean complianceAllows(ComplianceViolation violation, String detail);
+
+    <T extends Throwable> void complianceAssert(ComplianceViolation violation, String detail, Function<String, T> error) throws T;
```
javadoc

gregw replied:
Also, would this not be better as a method on HttpConfiguration?  When would a channel ever have a different impl?

joakime replied:
Each channel has a (potentially) unique set of cVL due cVL.initialize() requirement. 
The HttpConfiguration approach is the wrong place for most use cases with cVL.
It only really fits (and barely at that) for pre-channel violations like from HttpParser (but even that isn't really a concern anymore with the channel based approach).

joakime replied:
javadoc added

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -141,14 +144,82 @@ public HttpChannelState(ConnectionMetaData connectionMetaData)
+        // Always report violation to listeners
+        _complianceViolationListener.onComplianceViolation(new ComplianceViolation.Event(mode, violation, detail, allowed));
+
+        return allowed;
```
This method is substantially the same as the one above.  Factor out the mode selection maybe to HttpConfiguration itself

joakime replied:
Removed

gregw commented on jetty-core/jetty-compression/jetty-compression-server/src/main/java/org/eclipse/jetty/compression/server/CompressionHandler.java:
```diff
@@ -271,16 +271,13 @@ public boolean handle(Request request, Response response, Callback callback) thr
                     if (qualityCSV == null)
                     {
-                        HttpConfiguration httpConfiguration = request.getConnectionMetaData().getHttpConfiguration();
+                        HttpChannel httpChannel = HttpChannel.from(request);
```
I think it is more natural to report violations to the HttpConfiguration rather than the HttpChannel

joakime replied:
But that's the wrong place and doesn't take into account the cVL.initialize() / cVL.onRequestBegin() / cVL.onRequestEnd() requirements.
The HttpChannel is also the place where these events fire.
The individual violations (cVL.onComplianceViolation()) occur between these.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -797,14 +804,23 @@ public List<ComplianceViolation.Listener> getComplianceViolationListeners()
      */
     public void notifyViolation(ComplianceViolation violation, String details)
     {
+        ComplianceViolation.Mode mode;
```
The mode selection code is duplicated again.     Perhaps we should just have the following methods on HttpConfiguration:

```java

    public ComplianceViolation.Mode getComplianceViolationModeFor(ComplianceViolation violation)
    {
        if (violation instanceof UriCompliance.Violation)
            return getUriCompliance();
        else if (violation instanceof HttpCompliance.Violation)
            return getHttpCompliance();
        else if (violation instanceof MultiPartCompliance.Violation)
            return getMultiPartCompliance();
        else if (violation instanceof CookieCompliance.Violation)
            return getRequestCookieCompliance();
        else
            throw new UnsupportedOperationException("Unsupported ComplianceViolation type: " + violation.getClass().getName());
    }

    public <T extends Throwable> void assertAllowed(ComplianceViolation violation, String detail, Function<String, T> error) throws T
    {
        ComplianceViolation.Mode mode = getComplianceViolationModeFor(violation);
        boolean allowed = mode.allows(violation);

        List<ComplianceViolation.Listener> listeners = getComplianceViolationListeners();
        ComplianceViolation.Event event = new ComplianceViolation.Event(mode, violation, detail, allowed);
        Throwable throwable = null;
        for (ComplianceViolation.Listener listener : listeners)
        {
            try
            {
                listener.onComplianceViolation(event);
            }
            catch (Throwable t)
            {
                throwable = ExceptionUtil.combine(throwable, t);
            }
        }
        ExceptionUtil.ifExceptionThrowUnchecked(throwable);

        if (!allowed)
            throw error.apply(violation.getDescription());
    }

```

joakime replied:
Removed, no longer being used.

gregw reviewed (commented):
Looking better, but still work to do.

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/12806/compliance-violation-listener-cleanup

joakime committed:
Changes from review

joakime committed:
Changes from review

joakime committed:
Example of ComplianceViolationListener behavior in PR

joakime committed:
Fixing onRequestBegin duplication

joakime committed:
Improved ComplianceViolationListenerTest

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/12806/compliance-violation-listener-cleanup

joakime committed:
Fixing up testcases to assert all compliance events.

joakime committed:
Introduce HttpConfiguration.setNotifyForbiddenComplianceViolations

joakime committed:
Work around flaky behavior seen with jetty-client and ee11 bad URLs

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -141,14 +144,89 @@ public HttpChannelState(ConnectionMetaData connectionMetaData)
+     * @throws T Throwable if not allowed
+     */
+    @Override
+    public <T extends Throwable> void complianceAssert(ComplianceViolation violation, String detail, Function<String, T> error) throws T
```
I would prefer this method to be of the form:

```java
boolean processViolation(ComplianceViolation violation, String detail);

if (!processViolation(violation, detail))
    throw ...
```

As it is so in many other places.

joakime committed:
Performing ambiguous path tests against all envs

joakime committed:
Ambiguous Path Test in ee11

joakime committed:
Ambiguous Path Test in core

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceViolation.java:
```diff
@@ -82,10 +82,13 @@ interface Mode
+    /**
+     * @deprecated no replacement
+     */
+    @Deprecated(forRemoval = true, since = "12.1.5")
```
Update to 12.1.6

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceViolation.java:
```diff
@@ -101,13 +104,13 @@ static void notify(List<ComplianceViolation.Listener> listeners, ComplianceViola
     }
 
-    record Event(ComplianceViolation.Mode mode, ComplianceViolation violation, String details)
+    record Event(ComplianceViolation.Mode mode, ComplianceViolation violation, String details, boolean allowed)
```
Add the previous constructor for backwards compatibility.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceViolation.java:
```diff
@@ -200,7 +203,10 @@ private CapturingListener(List<Event> events)
         {
-            return new CapturingListener(new ArrayList<>());
+            List<Event> requestEvents = new ArrayList<>();
+            if (events != null)
```
`this.events` should be made null-safe, also because of how `onComplianceViolation()` is implemented below.

Then this code simplifies a lot.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceViolationException.java:
```diff
@@ -19,7 +19,7 @@ public class ComplianceViolationException extends IllegalArgumentException
     public ComplianceViolationException(ComplianceViolation.Mode mode, ComplianceViolation violation, String details)
     {
-        this(new ComplianceViolation.Event(mode, violation, details));
+        this(new ComplianceViolation.Event(mode, violation, details, false));
```
Use old constructor.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -381,6 +382,41 @@ public HttpCompliance without(String name, Violation... violations)
+        boolean allowed = this.allows(violation);
+
+        // Always report violation to listeners
+        listener.onComplianceViolation(new ComplianceViolation.Event(this, violation, detail, allowed));
```
Wrap into `try/catch` to avoid user-thrown exceptions.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -401,8 +437,14 @@ private static Set<Violation> copyOf(Set<Violation> violations)
+     * @param listener the notification method for violations.  (Tip: use the Request specific Listener from the {@code HttpChannelState})
+     * @throws HttpException.RuntimeException if there is a violation that wasn't allowed
+     */
+    public void check(MetaData.Request request, ComplianceViolation.Listener listener)
```
Rename to `notifyAndAssert()`

joakime replied:
Consider a new `ComplianceUtils.notifyAndAssert(HttpCompliance, MetaData.Request request, ComplianceViolation.Listener listener)` static.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -419,43 +461,43 @@ public static void checkHttpCompliance(MetaData.Request request, HttpCompliance
+    /**
+     * @deprecated use {@link #check(MetaData.Request, ComplianceViolation.Listener)} instead.
+     */
+    @Deprecated(forRemoval = true, since = "12.1.5")
```
Update to 12.1.6

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -150,9 +149,9 @@ static Mutable build(HttpFields fields, EnumSet<HttpHeader> removeFields)
     }
 
-    static Mutable build(HttpCompliance httpCompliance, BiConsumer<ComplianceViolation, String> notifyViolation)
+    static Mutable build(HttpCompliance httpCompliance, Supplier<ComplianceViolation.Listener> listenerSupplier)
```
`public` method, can't be changed. Deprecate and add the new one.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -150,9 +149,9 @@ static Mutable build(HttpFields fields, EnumSet<HttpHeader> removeFields)
+    static Mutable build(HttpCompliance httpCompliance, Supplier<ComplianceViolation.Listener> listenerSupplier)
     {
-        return new org.eclipse.jetty.http.MutableHttpFields.Compliant(httpCompliance, notifyViolation);
+        return new org.eclipse.jetty.http.MutableHttpFields.Compliant(httpCompliance, listenerSupplier);
```
No need of the package?

joakime replied:
Need package as there's 2 MutableHttpFields objects within this scope (see other existing methods that do this same pattern)

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartCompliance.java:
```diff
@@ -209,6 +210,29 @@ public Set<Violation> getAllowed()
+     * @param <T> the type of Throwable
+     * @throws T Throwable if not allowed
+     */
+    public <T extends Throwable> void assertAllowed(MultiPartCompliance.Violation violation, ComplianceViolation.Listener listener, String detail, Function<String, T> error) throws T
```
Rename to `notifyAndAssert()` or `reportAndAssert()`.
Not a big fan of this `public` method, we do not want users to call it, it's just called by the implementation and should be `private` somewhere else.

joakime replied:
Similar to HttpCompliance review

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -703,6 +703,14 @@ int getPartsSize()
             return listener.getPartsSize();
         }
 
+        private boolean complianceAllows(ComplianceViolation violation, String reason)
```
This method duplicates `MultiPartCompliace.assertAllowed()`. 
I would prefer this private method instead of a public method in `MultiPartCompliace`.

joakime replied:
Need to removing duplication.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -916,13 +907,14 @@ public void onViolation(MultiPartCompliance.Violation violation)
-                    ComplianceViolation.Event event = new ComplianceViolation.Event(compliance, violation, "multipart spec violation");
+                    boolean allowed = compliance.allows(violation);
+                    ComplianceViolation.Event event = new ComplianceViolation.Event(compliance, violation, "multipart spec violation", allowed);
                     complianceListener.onComplianceViolation(event);
```
Seems another duplicate of `complianceAllows()` above?

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -916,13 +907,14 @@ public void onViolation(MultiPartCompliance.Violation violation)
                 {
                     if (LOG.isDebugEnabled())
-                        LOG.atDebug().setCause(x).log("failure while notifying listener {}", complianceListener);
+                        LOG.debug("failure while notifying listener {}", complianceListener, x);
```
I prefer the previous style, now.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java:
```diff
@@ -450,10 +435,26 @@ else if (_complianceMode.allows(ATTRIBUTES))
+        };
+    }
+
+    private boolean complianceAllows(ComplianceViolation violation, String reason)
```
Rename to  `notifyAndCheck()` or `notifyAndVerify()`.

joakime replied:
Going to use `notifyAndCheck()`

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java:
```diff
@@ -450,10 +435,26 @@ else if (_complianceMode.allows(ATTRIBUTES))
+    private void complianceNotify(ComplianceViolation violation, String reason, boolean allowed)
+    {
+        if (_complianceListener != null)
+            _complianceListener.onComplianceViolation(new ComplianceViolation.Event(_complianceMode, violation, reason, allowed));
```
Wrap in `try/catch`.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -454,6 +459,64 @@ public static boolean isPathViolation(UriCompliance.Violation violation)
+     * @param <T> the type of Throwable
+     * @throws T Throwable if not allowed
+     */
+    public <T extends Throwable> void assertAllowed(UriCompliance.Violation violation, ComplianceViolation.Listener listener, String detail, Function<String, T> error) throws T
```
Not a fan of these public methods, should be hidden from users.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -474,18 +537,35 @@ private static Set<Violation> copyOf(Set<Violation> violations)
+     * @deprecated replaced with {@link #assertAllowed(HttpURI, ComplianceViolation.Listener, Function)}
+     * @see #assertAllowed(HttpURI, ComplianceViolation.Listener, Function)
+     */
+    @Deprecated(since = "12.1.5", forRemoval = true)
```
Update to 12.1.6.

sbordet commented on jetty-core/jetty-http3/jetty-http3-server/src/main/java/module-info.java:
```diff
@@ -18,6 +18,7 @@
     requires transitive org.eclipse.jetty.io;
     requires transitive org.eclipse.jetty.quic.server;
     requires transitive org.eclipse.jetty.server;
+    requires org.eclipse.jetty.http;
```
Remove, not necessary as it is required transitively from jetty-server.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -141,14 +144,89 @@ public HttpChannelState(ConnectionMetaData connectionMetaData)
+            _complianceViolationListener = new AllowedOnlyComplianceListener(_complianceViolationListener);
+    }
+
+    public static class AllowedOnlyComplianceListener implements ComplianceViolation.Listener
```
Can it be `private`?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1329,27 +1333,16 @@ public void parsedHeader(HttpField field)
 
         public Runnable headerComplete()
         {
-            UriCompliance compliance;
```
Is it ok to remove this block?

joakime replied:
Yes, it's handled elsewhere.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -131,6 +132,19 @@ default Runnable onClose()
+     * @param <T> the type of Throwable
+     * @throws T Throwable if not allowed
+     */
+    <T extends Throwable> void complianceAssert(ComplianceViolation violation, String detail, Function<String, T> error) throws T;
```
Rename to `notifyAndAssert()`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlParameterViolationListener.java:
```diff
@@ -0,0 +1,44 @@
+
+package org.eclipse.jetty.util;
+
+public interface UrlParameterViolationListener
```
Feels weird as we don't have such things for other violation listeners.
Why are not these additional UriCompliance.Violations, and we notify them there?

joakime replied:
This bridges `jetty-util` (which cannot see `jetty-http`) in a way that the violations can be reported properly.

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedCSV.java:
```diff
@@ -197,35 +196,47 @@ public String toString()
+                boolean allowed = _complianceMode.allows(violation);
+                _listener.onComplianceViolation(new ComplianceViolation.Event(_complianceMode, violation, value, allowed));
+                if (!allowed)
+                    throw new BadMessageException("Invalid quoted: " + value);
```
`BadMessageException` is deprecated and has been replaced by `HttpException.RuntimeException`.

joakime committed:
Changes from review

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/12806/compliance-violation-listener-cleanup

joakime committed:
Changes from review

joakime committed:
Removing UrlParameterViolationListener

lorban committed:
replace hardcoded value with constant

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
complete the stream with a failure only when consumeAvailable reports an exception or the last write is failed

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceUtils.java:
```diff
@@ -0,0 +1,187 @@
+     * @param complianceListener the listener to notify (can be null)
+     * @return true if violation is allowed per the configured compliance.
+     */
+    public static boolean allowed(ComplianceViolation.Mode compliance,
```
Rename to `notifyAndCheck()`

joakime replied:
The `notifyAndCheck` is bad name, it doesn't convey what the purpose is.
It's to check if allowed.
The `notify` is noise (as far as the method name is concerned), and doesn't belong in the method name.

joakime replied:
Find a better name, a single verb please.

After all we don't call our `Handler.handle()` method something like `Handle.canRequestAndProduceAResponse()`

joakime replied:
Also, the method takes a `ComplianceViolation.Listener` as a parameter, you don't need to include "notify" in the method name.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ComplianceUtils.java:
```diff
@@ -0,0 +1,187 @@
+        }
+    }
+
+    public static <T extends Throwable> void notifyAndAssert(ComplianceViolation.Mode compliance,
```
I'd prefer the `*Assert` methods to be removed, as they don't add anything, and they're as verbose as just calling `allowed()/notifyAndCheck()` plus throwing.

joakime replied:
so you want to go from ...

``` java
if (checkCompliance(compliance, violation1, detail, error))
{
   // do allowed part for violation1
} 
else if (checkCompliance(compliance, violation2, detail, error))
{
   // do allowed part for violation2
}
```

to

``` java
if (allowed(compliance, violation1))
{
   // do allowed part for violation1
} 
else 
{
   throw error(violation1, detail)
}

if (allowed(compliance, violation2))
{
  // do allowed part for violation2
} 
else 
{
  throw error(violation2, detail)
}
```

That's going to blossom the lines of code a significant bit

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -381,6 +381,48 @@ public HttpCompliance without(String name, Violation... violations)
+     * @param <T> the type of Throwable
+     * @throws T Throwable if not allowed
+     */
+    public <T extends Throwable> void assertAllowed(HttpCompliance.Violation violation, ComplianceViolation.Listener listener, Function<String, T> error) throws T
```
Move to ComplianceUtils.

joakime replied:
Removed

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java:
```diff
@@ -381,6 +381,48 @@ public HttpCompliance without(String name, Violation... violations)
+     * @param <T> the type of Throwable
+     * @throws T Throwable if not allowed
+     */
+    public <T extends Throwable> void assertAllowed(HttpCompliance.Violation violation, ComplianceViolation.Listener listener, String detail, Function<String, T> error) throws T
```
Ditto.

joakime replied:
Removed

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java:
```diff
@@ -914,15 +899,17 @@ public void onFailure(Throwable failure)
-                    ComplianceViolation.Event event = new ComplianceViolation.Event(compliance, violation, "multipart spec violation");
+                    boolean allowed = compliance.allows(violation);
+                    ComplianceViolation.Event event = new ComplianceViolation.Event(compliance, violation, "multipart spec violation", allowed);
                     complianceListener.onComplianceViolation(event);
```
This is done twice now?

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/QuotedQualityCSV.java:
```diff
@@ -330,36 +329,43 @@ public String toString()
+            boolean allowed = _complianceMode.allows(violation);
+            _listener.onComplianceViolation(new ComplianceViolation.Event(_complianceMode, violation, value, allowed));
+            if (!allowed)
+                throw new HttpException.RuntimeException(HttpStatus.BAD_REQUEST_400, "Invalid quoted-quality: " + value);
```
Use ComplianceUtils.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java:
```diff
@@ -86,7 +79,7 @@ public void parseField(String field)
             {
-                if (!_complianceMode.allows(INVALID_COOKIES))
+                if (!ComplianceUtils.allowed(_complianceMode, CookieCompliance.Violation.INVALID_COOKIES, field, _complianceListener))
                      throw new InvalidCookieException("Invalid Cookie character: 0x%02x [%s]: %s".formatted((byte)c, c, field));
```
I like this style much better than `ComplianceUtils.notifyAndAssert()`.

joakime replied:
This is the most simplistic use case of compliance checking too.
It's a negative check, with no allow logic.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java:
```diff
@@ -284,51 +269,51 @@ else if (_complianceMode.allows(INVALID_COOKIES))
                     {
-                        string.append(c);
-                        if (!cookieInvalid)
+                        if (ComplianceUtils.allowed(_complianceMode, CookieCompliance.Violation.INVALID_COOKIES, "Illegal character '%s' in quoted section in %s".formatted(c, field), _complianceListener))
```
Merge into `else if` as before.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/RFC6265CookieParser.java:
```diff
@@ -377,10 +359,12 @@ else if (_complianceMode.allows(INVALID_COOKIES))
                     {
-                        reportComplianceViolation(OPTIONAL_WHITE_SPACE, field);
-                        continue;
+                        if ((c == ' ' || c == '\t') && ComplianceUtils.allowed(_complianceMode, CookieCompliance.Violation.OPTIONAL_WHITE_SPACE, field, _complianceListener))
```
Merge into `else if` as before

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -474,18 +479,34 @@ private static Set<Violation> copyOf(Set<Violation> violations)
-                    if (listener != null)
-                        listener.onComplianceViolation(new ComplianceViolation.Event(compliance, violation, uri.toString()));
+                boolean allowed = compliance.allows(violation);
+
+                // Always report violation to listeners
+                if (listener != null)
+                    listener.onComplianceViolation(new ComplianceViolation.Event(compliance, violation, uri.toString(), allowed));
 
+                // Only trigger a failure of the HttpURI for compliance reasons if the compliance doesn't allow for violation detected
+                if (!allowed)
```
Use ComplianceUtils.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -751,9 +809,9 @@ public void run()
-                    LOG.debug("completeStream({}, {})", stream, Objects.toString(failure));
-                completeStream(stream, failure);
+                if (completeStreamFailure == null)
+                    completeStreamFailure = request._callback.unconsumed();
```
I don't like this call be outside of the locked block above: the failure should be retrieved there into a local variable that is then used here.

If done inside the locked block, you can remove the method and access directly the field.

joakime replied:
@lorban this is your new block of code.

lorban replied:
The `_unconsumed` field was written with the `ChannelRequest` lock, which is different from the `HttpChannelState` one taken above. This is why I added this little helper getter.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -788,16 +856,26 @@ private void completeLastWrite(Throwable failure)
-                completeStream(stream, failure);
+            {
+                if (completeStreamFailure == null)
+                    completeStreamFailure = channelCallback.unconsumed();
```
Ditto.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1676,15 +1764,28 @@ else if (!httpChannelState.lockedIsLastStreamSendCompleted())
             else if (completeStream)
-                httpChannelState.completeStream(stream, failure);
+            {
+                Throwable x = httpChannelState._lastWriteCallback.failure();
```
Ditto, I prefer the exception to retrieved in the locked block above.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1676,15 +1764,28 @@ else if (!httpChannelState.lockedIsLastStreamSendCompleted())
+            {
+                Throwable x = httpChannelState._lastWriteCallback.failure();
+                if (completeStreamFailure == null || x != null)
+                    completeStreamFailure = x;
```
Combine exceptions instead of ignoring `x` when `completeStreamFailure != null`?

joakime replied:
@lorban this is yours too

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1390,7 +1388,7 @@ public boolean is100ContinueExpected()
 
             Request request = _httpChannel.getRequest();
-            getHttpChannel().getComplianceViolationListener().onRequestBegin(request);
+            _httpChannel.getComplianceViolationListener().onRequestBegin(request);
```
```suggestion
            getComplianceViolationListener().onRequestBegin(request);
```

joakime replied:
Invalid suggestion, `getComplianceViolationListener()` is not a method on `HttpConnection`
That method exists as part of the `RequestHandler implements HttpParser.RequestHandler`

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1329,27 +1334,20 @@ public void parsedHeader(HttpField field)
-                        else
-                            throw new HttpException.RuntimeException(HttpStatus.BAD_REQUEST_400, "Authority!=Host");
+                        HttpCompliance httpCompliance = getHttpChannel().getHttpConfiguration().getHttpCompliance();
+                        ComplianceViolation.Listener complianceListener = getHttpChannel().getComplianceViolationListener();
```
```suggestion
                        ComplianceViolation.Listener complianceListener = getComplianceViolationListener();
```

joakime replied:
Not a valid suggestion, method doesn't exist here.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1138,10 +1138,16 @@ public boolean contentComplete()
         public void onViolation(ComplianceViolation.Event event)
         {
-            getHttpChannel().getComplianceViolationListener().onComplianceViolation(event);
+            getComplianceViolationListener().onComplianceViolation(event);
```
Call `ComplianceUtils.notify()`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -130,7 +130,11 @@ default Runnable onClose()
      */
     void initialize();
 
+    HttpConfiguration getHttpConfiguration();
```
This is unnecessary, remove.
In all places where this method is used, the `HttpConfiguration` is available via calls to reachable components.

joakime replied:
There is the `CompressionHandler` and the `HttpConnection.HttpStreamOverHTTP1` that needs this API.

Eg: The `CompressionHandler` would need to know the active `HttpChannel` to know the active `Connector` to then lookup the active `ConnectionFactory` to then lookup the in use `HttpConfiguration`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlParameterDecoder.java:
```diff
@@ -36,6 +36,9 @@ class UrlParameterDecoder
     private final boolean allowTruncatedEncoding;
     private final CharsetStringBuilder builder;
+    private final BiConsumer<String, Boolean> onBadEncoding;
+    private final BiConsumer<String, Boolean> onBadPercentage;
```
Rename to `onBadPercent`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlParameterDecoder.java:
```diff
@@ -63,9 +66,15 @@ public UrlParameterDecoder(CharsetStringBuilder charsetStringBuilder, BiConsumer
      * @param allowBadPercent allow use of bad pct-encoding with the {@link CharsetStringBuilder} (optional behavior)
      * @param allowTruncatedEncoding allow use of truncated pct-encoding with the {@link CharsetStringBuilder} (optional behavior)
+     * @param onBadEncoding consumer for bad encoding events
+     * @param onBadPercentage consumer for bad percentage events
```
Rename to `onBadPercent`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/UrlParameterDecoder.java:
```diff
@@ -63,9 +66,15 @@ public UrlParameterDecoder(CharsetStringBuilder charsetStringBuilder, BiConsumer
-                               boolean allowBadEncoding, boolean allowBadPercent, boolean allowTruncatedEncoding)
+                               boolean allowBadEncoding, boolean allowBadPercent, boolean allowTruncatedEncoding,
+                               BiConsumer<String, Boolean> onBadEncoding,
+                               BiConsumer<String, Boolean> onBadPercentage,
```
Rename to `onBadPercent`.

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/12806/compliance-violation-listener-cleanup

joakime committed:
Merge branch 'fix/12.1.x/12806/compliance-violation-listener-cleanup' of github.com:jetty/jetty.project into fix/12.1.x/12806/compliance-violation-listener-cleanup

joakime committed:
Pass through for code review

joakime committed:
Removing recently added HttpChannel.getHttpConfiguration()

joakime committed:
Changes from review of naming

joakime committed:
Fix missed BadMessageException

joakime committed:
Fix testcase (duplicate notifications fixed in prior commit)

lorban committed:
fix stream completion on error

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
update log

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
1st shot

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
consumeAvailableFailure moved

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
lastWriteCallback._failure moved

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
_writeFailure moved

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
check _callbackFailure on FAILED state

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
todo

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove todo

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Improved error handling when the response failed to be generated, but the stream was not committed yet: now we call the error handling mechanism.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
In case of write failures, the write callback does not store the failure anymore.
The reason is that the failure is notified either to an application callback (which should eventually notify the handler callback), or to one of our implementation callbacks, which will do the proper action, including storing the failure if necessary.
A special case is when the failure triggers the ErrorHandler, which may produce a successful response.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Re-added null guard to ChannelResponse.succeeded|failed because the write callback can be stolen by an asynchronous failure.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed test now that behavior for HTTP/1.1 is different with respect to invoking the ErrorHandler for response generation failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Disabling infinispan tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Upon completing the last write, set `state = StreamSendState.FAILED` if there was a failure.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed ServerFCGIConnection buffer leak in case of reading the terminal chunk.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
Fixed test's assertions to make them check that the stream is failed when no response is generated

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Added comment about HTTP/3 failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Covered more cases.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
fix test to make it use onComplete instead of onResponseSuccess as we expect the stream to be failed by the server

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Combined exceptions in lockedCompleteStreamFailure().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merge branch 'fix/12.1.x/channel-state-failure-handling' into fix/12.1.x/12806/compliance-violation-listener-cleanup.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/12806/compliance-violation-listener-cleanup
